src/OrleansAzureUtils/Providers/Storage/AzureTableStorage.cs

```diff
@@ -246,49 +250,112 @@ internal void ConvertToStorageFormat(object grainState, GrainStateEntity entity)
 
                 if (Log.IsVerbose3) Log.Verbose3("Writing binary data size = {0} for grain id = Partition={1} / Row={2}",
                     data.Length, entity.PartitionKey, entity.RowKey);
-                
-                dataSize = data.Length;
-                entity.Data = data;
+
+                CheckMaxDataSize(data.Length, MAX_DATA_CHUNK_SIZE * MAX_DATA_CHUNKS_COUNT);
+
+                foreach (var partition in SplitBinaryData(data).Zip(GetBinaryDataPropertyNames(), (chunk, name) => new { chunk, name }))
+                {
+                    entity.Properties.Add(partition.name, new EntityProperty(partition.chunk));
+                }
             }
-            if (dataSize > MAX_DATA_SIZE)
+        }
+
+        private void CheckMaxDataSize(int dataSize, int maxDataSize)
+        {
+            if (dataSize > maxDataSize)
             {
-                var msg = string.Format("Data too large to write to Azure table. Size={0} MaxSize={1}", dataSize, MAX_DATA_SIZE);
+                var msg = string.Format("Data too large to write to Azure table. Size={0} MaxSize={1}", dataSize, maxDataSize);
                 Log.Error(0, msg);
                 throw new ArgumentOutOfRangeException("GrainState.Size", msg);
             }
         }
 
+        private static IEnumerable<byte[]> SplitBinaryData(byte[] binaryData)
+        {
+            var chunkCount = (binaryData.Length - 1) / MAX_DATA_CHUNK_SIZE + 1;
+            for (var i = 0; i < chunkCount; i++)
+            {
+                yield return binaryData.Skip(i * MAX_DATA_CHUNK_SIZE).Take(MAX_DATA_CHUNK_SIZE).ToArray();
```

Comment:
If we're moving away from the skip/take pattern, it may be valuable just to carve the buffer up using ArraySegments, then converting the segments into arrays.  Would make code simple an clean.  Just a suggestion.